apis/openstack-nova/src/main/java/org/jclouds/openstack/nova/v2_0/compute/functions/AllocateAndAddFloatingIpToNode.java

```diff
@@ -225,28 +228,64 @@ public boolean apply(@Nullable org.jclouds.openstack.neutron.v2.domain.FloatingI
       return floatingIPOptional;
    }
 
-   private org.jclouds.openstack.neutron.v2.domain.FloatingIP createFloatingIpUsingNeutron(org.jclouds.openstack.neutron.v2.features.FloatingIPApi neutronFloatingApi, NodeMetadata node, final String availabilityZone) {
+   private org.jclouds.openstack.neutron.v2.domain.FloatingIP createFloatingIpUsingNeutron(org.jclouds.openstack.neutron.v2.features.FloatingIPApi neutronFloatingApi,
+                                                                                           NodeMetadata node, Optional<Set<String>> poolNames, final String availabilityZone) {
       String regionId = node.getLocation().getParent().getId();
 
-      Optional<Network> networkOptional = getNetworkApi(regionId).list().concat().firstMatch(new Predicate<Network>() {
-         @Override
-         public boolean apply(@Nullable Network input) {
-            return input.getExternal() && input.getAvailabilityZone().equals(availabilityZone);
+      List<Network> networks = getNetworksFromPoolNames(regionId, poolNames);
+      // 1.) Attempt to allocate from optionally passed poolNames
+      org.jclouds.openstack.neutron.v2.domain.FloatingIP floatingIP = null;
+      for (Network network : networks) {
+         try {
+            logger.debug(">> allocating floating IP from network %s for node(%s)", network, node);
+            org.jclouds.openstack.neutron.v2.domain.FloatingIP createFloatingIP = org.jclouds.openstack.neutron.v2.domain.FloatingIP.CreateFloatingIP.createBuilder(network.getId()).availabilityZone(network.getAvailabilityZone()).build();
+            floatingIP = neutronFloatingApi.create((org.jclouds.openstack.neutron.v2.domain.FloatingIP.CreateFloatingIP) createFloatingIP);
+            logger.debug(">> allocated floating IP(%s) from network(%s) for node(%s)", floatingIP, network, node);
+            break;
+         } catch (Exception ex) {
+            logger.trace("<< [%s] failed to allocate a floating IP from network %s for node(%s)", ex.getMessage(), network, node);
          }
-      });
-      if (!networkOptional.isPresent())
-         throw new InsufficientResourcesException("Failed to find a suitable external network.");
+      }
 
-      org.jclouds.openstack.neutron.v2.domain.FloatingIP createFloatingIP = org.jclouds.openstack.neutron.v2.domain.FloatingIP.CreateFloatingIP
-              .createBuilder(networkOptional.get().getId())
-              .availabilityZone(networkOptional.get().getAvailabilityZone())
-              .build();
+      // 2.) Attempt to allocate from a external network
+      if (floatingIP == null) {
+         Optional<Network> networkOptional = Iterables.tryFind(networks, new Predicate<Network>() {
+            @Override
+            public boolean apply(Network input) {
+               return availabilityZone.equals(input.getAvailabilityZone()) && input.getExternal();
+            }
+         });
```

Comment:
If the pool names is a list of more than one element, we try to allocate the floating IP in one of the given networks. Why not apply the same pattern if no network is defined? We can try every external network in the same availability zone, following the exact same pattern.